File: agent/agent.py
#!/usr/bin/env python3
"""
Gemini Nano Browser Agent v2
Uses agent-browser (Vercel Labs) for browser control + Gemini Nano for brain.

Agent gets its OWN Chromium instance — no need to touch your main Chrome.

Usage:
    python3 agent.py "Go to github.com and find trending Python repos"
    python3 agent.py --interactive
"""

import asyncio
import json
import sys
import os
import subprocess
import base64
from pathlib import Path

try:
    import httpx
except ImportError:
    print("Installing httpx...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "httpx"])
    import httpx


# ══════════════════════════════════════════════════════════
# Configuration
# ══════════════════════════════════════════════════════════

BRIDGE_URL = "http://localhost:8765"
AB_CMD = "agent-browser"  # CLI command

AGENT_SYSTEM = """You are a browser automation agent. You control a real Chromium browser via agent-browser commands.

CAPABILITIES:
- Navigate to URLs
- Click elements (by ref @e1, CSS selector, or text)
- Type/fill text into input fields
- Scroll the page
- Take screenshots and analyze them
- Read page content (accessibility tree)
- Submit forms, press keys
- Wait for page loads

HOW TO INTERACT:
You see the page as an accessibility tree with numbered refs (@e1, @e2, ...).
Each ref corresponds to a clickable/interactive element.
Use refs for precise element targeting.

OUTPUT FORMAT:
Always respond with ONE JSON action block:

{
    "thought": "What I observe and what I plan to do",
    "action": "navigate|click|fill|type|press|scroll|screenshot|snapshot|read|wait|done",
    "params": { ... }
}

ACTIONS:
- navigate:  {"url": "https://..."}
- click:     {"ref": "@e1"} or {"text": "Sign In"} or {"selector": "button.submit"}
- fill:      {"ref": "@e3", "text": "hello@example.com"}
- type:      {"ref": "@e3", "text": "hello world"}
- press:     {"key": "Enter"} or {"key": "Tab"}
- scroll:    {"direction": "down", "amount": 500}
- screenshot: {}  — capture page screenshot
- snapshot:  {}  — get accessibility tree (refs)
- read:      {}  — get visible text content
- wait:      {"seconds": 2}
- done:      {"result": "task completion summary"}

RULES:
1. Always start by taking a snapshot to see what's on the page
2. Use refs (@e1, @e2) when available — they're reliable
3. If an action fails, try a different approach
4. Max 25 steps — complete the task efficiently
5. When done, use the "done" action"""


# ══════════════════════════════════════════════════════════
# Gemini Nano Bridge Client
# ══════════════════════════════════════════════════════════

class GeminiBridge:
    def __init__(self):
        self.http = httpx.AsyncClient(base_url=BRIDGE_URL, timeout=120)

    async def chat(self, messages: list) -> str:
        resp = await self.http.post("/v1/chat/completions", json={
            "messages": messages,
            "temperature": 0.7,
        })
        data = resp.json()
        if "error" in data:
            raise Exception(f"Bridge error: {data['error']}")
        return data["choices"][0]["message"]["content"]

    async def analyze_image(self, image_b64: str, prompt: str) -> str:
        resp = await self.http.post("/v1/analyze-image", json={
            "image": image_b64,
            "prompt": prompt,
        })
        data = resp.json()
        return data.get("result", data.get("error", "No result"))

    async def health(self) -> bool:
        try:
            resp = await self.http.get("/health")
            return resp.json().get("status") == "ok"
        except Exception:
            return False


# ══════════════════════════════════════════════════════════
# Agent-Browser Wrapper
# ══════════════════════════════════════════════════════════

class AgentBrowser:
    """Wrap agent-browser CLI for the agent."""

    def __init__(self, session: str = "gemini-agent"):
        self.session = session
        self.processed = 0

    def run(self, *args, timeout: int = 30) -> dict:
        """Run an agent-browser command and return JSON result."""
        cmd = [AB_CMD, "--session", self.session, "--json"] + list(args)
        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            if result.stdout.strip():
                return json.loads(result.stdout.strip())
            return {"success": False, "error": result.stderr.strip() or "No output"}
        except subprocess.TimeoutExpired:
            return {"success": False, "error": "Command timed out"}
        except json.JSONDecodeError as e:
            return {"success": False, "error": f"JSON parse error: {e}"}
        except Exception as e:
            return {"success": False, "error": str(e)}

    def open(self, url: str) -> dict:
        return self.run("open", url)

    def close(self) -> dict:
        return self.run("close")

    def snapshot(self) -> str:
        """Get accessibility tree with refs."""
        result = self.run("snapshot", "-i")
        if result.get("success"):
            return json.dumps(result.get("data", {}), indent=2)
        return f"Snapshot error: {result.get('error', 'unknown')}"

    def screenshot(self, path: str = "/tmp/agent-screenshot.png") -> dict:
        return self.run("screenshot", path, "--annotate")

    def click(self, target: str) -> dict:
        return self.run("click", target)

    def fill(self, ref: str, text: str) -> dict:
        return self.run("fill", ref, text)

    def type_text(self, ref: str, text: str) -> dict:
        return self.run("type", ref, text)

    def press(self, key: str) -> dict:
        return self.run("press", key)

    def scroll(self, direction: str = "down", amount: int = 500) -> dict:
        return self.run("scroll", direction, str(amount))

    def get_text(self, ref: str = None) -> str:
        if ref:
            result = self.run("get", "text", ref)
        else:
            # Get all text from page
            result = self.run("eval", "document.body.innerText.substring(0, 8000)")
        if result.get("success"):
            return result.get("data", {}).get("text", "") or str(result.get("data", ""))
        return f"Get text error: {result.get('error', 'unknown')}"

    def get_title(self) -> str:
        result = self.run("get", "title")
        if result.get("success"):
            return result.get("data", {}).get("title", "")
        return ""

    def get_url(self) -> str:
        result = self.run("get", "url")
        if result.get("success"):
            return result.get("data", {}).get("url", "")
        return ""

    def screenshot_base64(self) -> str:
        """Take screenshot and return base64."""
        path = f"/tmp/agent-screenshot-{self.processed}.png"
        self.processed += 1
        result = self.screenshot(path)
        if result.get("success") and os.path.exists(path):
            with open(path, "rb") as f:
                return base64.b64encode(f.read()).decode()
        return ""


# ══════════════════════════════════════════════════════════
# Agent Loop
# ══════════════════════════════════════════════════════════

class BrowserAgent:
    def __init__(self):
        self.gemini = GeminiBridge()
        self.browser = AgentBrowser()
        self.history = []
        self.max_steps = 25

    async def initialize(self):
        print("=" * 60)
        print("  Gemini Nano Browser Agent v2")
        print("  (agent-browser + Gemini Nano)")
        print("=" * 60)

        # Check bridge
        print("\n[1/3] Checking Gemini Nano bridge...")
        if await self.gemini.health():
            print("  ✅ Bridge connected")
        else:
            print("  ❌ Bridge not running!")
            print(f"     Start: cd /home/bobby/gemini-nano-demo/bridge && node server.js")
            return False

        # Check agent-browser
        print("[2/3] Checking agent-browser...")
        result = self.browser.run("--version")
        if result.get("success") or "version" in str(result.get("data", "")).lower():
            print("  ✅ agent-browser installed")
        else:
            # Try getting version differently
            try:
                r = subprocess.run([AB_CMD, "--version"], capture_output=True, text=True, timeout=5)
                if r.returncode == 0:
                    print(f"  ✅ agent-browser: {r.stdout.strip()}")
                else:
                    print("  ❌ agent-browser not found")
                    print("     Install: npm install -g agent-browser && agent-browser install")
                    return False
            except Exception:
                print("  ❌ agent-browser not working")
                return False

        # Launch browser
        print("[3/3] Launching Chromium instance...")
        result = self.browser.open("about:blank")
        if result.get("success"):
            print("  ✅ Chromium launched (own instance, not your Chrome)")
        else:
            print(f"  ❌ Failed to launch: {result.get('error')}")
            return False

        print("\n" + "=" * 60)
        return True

    async def run_task(self, task: str):
        print(f"\n📋 Task: {task}")
        print("-" * 60)

        self.history = [
            {"role": "system", "content": AGENT_SYSTEM},
            {"role": "user", "content": f"Task: {task}\n\nStart by taking a snapshot to see the current page state."},
        ]

        for step in range(self.max_steps):
            print(f"\n--- Step {step + 1}/{self.max_steps} ---")

            # Get Gemini Nano's action (with retry on bridge errors)
            response = None
            for retry in range(3):
                try:
                    response = await self.gemini.chat(self.history)
                    break
                except Exception as e:
                    print(f"  ⚠️ Bridge error (attempt {retry+1}/3): {e}")
                    if retry < 2:
                        await asyncio.sleep(3)
                    else:
                        print("  ❌ Bridge not responding after 3 attempts")
                        return "Bridge connection lost"

            # Show thought
            thought = self.extract_thought(response)
            if thought:
                print(f"🧠 {thought[:200]}")

            # Parse action
            action = self.parse_action(response)

            if not action:
                self.history.append({"role": "assistant", "content": response})
                self.history.append({"role": "user", "content": "Please respond with a JSON action block."})
                continue

            act = action.get("action", "")
            params = action.get("params", {})
            print(f"⚡ {act}({params})")

            # Execute action
            result = await self.execute_action(action)
            print(f"📋 {result[:200]}")

            # Feed result back
            self.history.append({"role": "assistant", "content": response})
            self.history.append({"role": "user", "content": f"Result: {result}"})

            if act == "done":
                print(f"\n✅ Complete: {action.get('params', {}).get('result', 'Done')}")
                return action.get("params", {}).get("result", "Done")

        print(f"\n⚠️ Max steps ({self.max_steps}) reached")
        return "Max steps reached"

    def extract_thought(self, response: str) -> str:
        import re
        match = re.search(r'"thought"\s*:\s*"([^"]+)"', response)
        return match.group(1) if match else ""

    def parse_action(self, response: str) -> dict:
        import re
        # Try code block first (most reliable)
        match = re.search(r'```(?:json)?\s*(\{.*?\})\s*```', response, re.DOTALL)
        if match:
            try:
                return json.loads(match.group(1))
            except json.JSONDecodeError:
                pass
        # Try to find JSON with "action" key — handle nested braces
        depth = 0
        start = None
        for i, ch in enumerate(response):
            if ch == '{':
                if depth == 0:
                    start = i
                depth += 1
            elif ch == '}':
                depth -= 1
                if depth == 0 and start is not None:
                    candidate = response[start:i+1]
                    if '"action"' in candidate:
                        try:
                            return json.loads(candidate)
                        except json.JSONDecodeError:
                            pass
                    start = None
        return None

    async def execute_action(self, action: dict) -> str:
        act = action.get("action", "")
        params = action.get("params", {})

        try:
            if act == "navigate":
                result = self.browser.open(params.get("url", ""))
                if result.get("success"):
                    await asyncio.sleep(2)
                    return f"Navigated to {params.get('url')}"
                return f"Navigation failed: {result.get('error')}"

            elif act == "click":
                if "ref" in params:
                    result = self.browser.click(params["ref"])
                elif "text" in params:
                    result = self.browser.click(f"text={params['text']}")
                elif "selector" in params:
                    result = self.browser.click(params["selector"])
                else:
                    return "No click target"
                return "Clicked" if result.get("success") else f"Click failed: {result.get('error')}"

            elif act == "fill":
                result = self.browser.fill(params.get("ref", ""), params.get("text", ""))
                return "Filled" if result.get("success") else f"Fill failed: {result.get('error')}"

            elif act == "type":
                result = self.browser.type_text(params.get("ref", ""), params.get("text", ""))
                return "Typed" if result.get("success") else f"Type failed: {result.get('error')}"

            elif act == "press":
                result = self.browser.press(params.get("key", "Enter"))
                return f"Pressed {params.get('key')}" if result.get("success") else f"Press failed"

            elif act == "scroll":
                result = self.browser.scroll(
                    params.get("direction", "down"),
                    params.get("amount", 500)
                )
                return "Scrolled" if result.get("success") else f"Scroll failed"

            elif act == "screenshot":
                b64 = self.browser.screenshot_base64()
                if b64:
                    analysis = await self.gemini.analyze_image(
                        b64,
                        "Describe this webpage screenshot in detail. What elements are visible? What can be clicked or interacted with? What text is shown?"
                    )
                    return f"Screenshot: {analysis[:2000]}"
                return "Screenshot failed"

            elif act == "snapshot":
                return self.browser.snapshot()[:4000]

            elif act == "read":
                return self.browser.get_text()[:4000]

            elif act == "wait":
                await asyncio.sleep(params.get("seconds", 2))
                return "Waited"

            elif act == "done":
                return params.get("result", "Task complete")

            else:
                return f"Unknown action: {act}"

        except Exception as e:
            return f"Error: {e}"

    async def close(self):
        self.browser.close()
        print("[Agent] Browser closed")


# ══════════════════════════════════════════════════════════
# CLI
# ══════════════════════════════════════════════════════════

async def main():
    agent = BrowserAgent()

    if not await agent.initialize():
        sys.exit(1)

    if len(sys.argv) > 1 and sys.argv[1] == "--interactive":
        print("\n🤖 Interactive mode. Type your task or 'quit' to exit.\n")
        while True:
            task = input("Task> ").strip()
            if task in ("quit", "exit", "q"):
                break
            if task:
                await agent.run_task(task)
    elif len(sys.argv) > 1:
        task = " ".join(sys.argv[1:])
        await agent.run_task(task)
    else:
        print("\nUsage:")
        print("  python3 agent.py 'Go to github.com and find trending Python repos'")
        print("  python3 agent.py --interactive")
        sys.exit(1)

    await agent.close()


if __name__ == "__main__":
    asyncio.run(main())

File: agent/agent.js
#!/usr/bin/env node
/**
 * Gemini Nano Browser Agent (Node.js)
 * Autonomous browser agent powered by Chrome's on-device Gemini Nano.
 * Uses agent-browser for browser control + Gemini Nano bridge for reasoning.
 *
 * Usage:
 *   node agent.js "Find AI news on Hacker News"
 *   node agent.js "Search Wikipedia for quantum computing" --headed
 *   node agent.js "Fill out contact form on example.com" --max-steps 15
 */

import { execFile } from 'child_process';
import http from 'http';

const AGENT_BROWSER = process.env.AGENT_BROWSER_BIN ||
  '/home/bobby/.local/lib/node_modules/agent-browser/bin/agent-browser.js';
const BRIDGE_URL = 'http://localhost:8765/v1/chat/completions';
const SESSION = 'gemini-nano-agent';
const DEFAULT_MAX_STEPS = 20;
const REQUEST_TIMEOUT = 120000; // 2 minutes for chat completions
const ACTION_TIMEOUT = 60000;   // 1 minute for agent-browser actions

const SYSTEM_PROMPT = `You are a browser automation agent. You control a web browser via accessibility tree snapshots AND WebMCP tools AND vision (screenshots).

Your job: complete the user's task by deciding the next action.

AVAILABLE ACTIONS (respond with JSON only):
- {"action": "click", "ref": "@eN"}        — Click element by accessibility ref
- {"action": "fill", "ref": "@eN", "value": "text"} — Clear and fill an input
- {"action": "type", "ref": "@eN", "value": "text"}  — Type into input (append)
- {"action": "press", "key": "Enter"}       — Press a keyboard key
- {"action": "scroll", "direction": "down", "amount": 500} — Scroll page
- {"action": "navigate", "url": "https://..."} — Go to URL
- {"action": "webmcp", "tool": "name", "args": {...}} — Call a WebMCP tool on the page
- {"action": "screenshot", "annotate": true}   — Take annotated screenshot (returns base64 PNG)
- {"action": "vision", "prompt": "describe...", "image": "base64"} — Analyze image with Gemini Nano (provide base64 from screenshot action)
- {"action": "done", "result": "summary"}   — Task is complete
- {"action": "stuck", "reason": "why"}      — Cannot proceed

RULES:
1. Always respond with valid JSON matching one of the actions above
2. Use refs from the accessibility tree (e.g., @e3, @e12)
3. WebMCP tools are semantic page capabilities — prefer them over raw clicks when available
4. Common WebMCP tools: search_page, extract_links, extract_forms, get_page_metadata, click_element, fill_form, scroll_to
5. VISION: Use "screenshot" to capture annotated image, then "vision" with prompt + that image to analyze visually
6. After each action, you'll see the updated page state
7. If a page needs loading, just say done for this step
8. If you can't find the right element, try scrolling
9. Be concise — just the JSON action, no explanation`;

function runAB(...args) {
  return new Promise((resolve, reject) => {
    const headed = args.includes('--headed');
    const cleanArgs = args.filter(a => a !== '--headed');
    const cmdArgs = [AGENT_BROWSER, '--session', SESSION, '--json', ...cleanArgs.map(String)];
    if (headed) cmdArgs.push('--headed');

    const child = execFile(AGENT_BROWSER, ['--session', SESSION, '--json', ...cleanArgs], {
      timeout: 60000,
      encoding: 'utf-8',
      maxBuffer: 1024 * 1024 * 10, // 10MB
    }, (error, stdout, stderr) => {
      if (error) {
        if (stdout) {
          try { return resolve(JSON.parse(stdout)); } catch { return resolve({ raw: stdout.trim() }); }
        }
        return reject(new Error(stderr?.trim() || error.message));
      }
      if (stdout.trim()) {
        try { resolve(JSON.parse(stdout)); }
        catch { resolve({ raw: stdout.trim() }); }
      } else {
        resolve({ ok: true });
      }
    });
  });
}

function getSnapshot(headed = false) {
  return runAB('snapshot', ...(headed ? ['--headed'] : []));
}

function safeParse(str) {
  try { return JSON.parse(str); } catch { return null; }
}

function geminiChat(messages, temperature = 0.3) {
  const payload = JSON.stringify({
    model: 'gemini-nano',
    messages,
    temperature,
    max_tokens: 256,
  });

  return new Promise((resolve) => {
    const url = new URL('http://localhost:8765/v1/chat/completions');
    const req = http.request({
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      timeout: 120000,
    }, (res) => {
      let body = '';
      res.on('data', c => body += c);
      res.on('end', () => {
        try {
          const data = JSON.parse(body);
          resolve(data.choices?.[0]?.message?.content?.trim() || '');
        } catch {
          resolve('{"action": "stuck", "reason": "Invalid bridge response"}');
        }
      });
    });
    req.on('error', e => resolve(`{"action": "stuck", "reason": "Bridge error: ${e.message}"}`));
    req.on('timeout', () => { req.destroy(); resolve('{"action": "stuck", "reason": "Bridge timeout"}'); });
    req.write(JSON.stringify({ model: 'gemini-nano', messages, temperature: 0.3, max_tokens: 256 }));
    req.end();
  });
}

function parseAction(text) {
  let t = text.trim();
  if (t.startsWith('```')) {
    t = t.split('\n').filter(l => !l.startsWith('```')).join('\n').trim();
  }
  const start = t.indexOf('{');
  const end = t.lastIndexOf('}') + 1;
  if (start >= 0 && end > start) {
    try { return JSON.parse(t.slice(start, end)); } catch {}
  }
  return { action: 'stuck', reason: `Could not parse: ${t.slice(0, 200)}` };
}

async function sleep(ms) {
  await new Promise(r => setTimeout(r, ms));
}

async function executeAction(action, headed = false) {
  const act = action.action;
  const asyncCases = ['screenshot', 'vision'];
  if (asyncCases.includes(act)) {
    return executeActionAsync(action, headed);
  }

  try {
    switch (action.action) {
      case 'click':
        await runAB('click', action.ref, ...(headed ? ['--headed'] : []));
        await sleep(1000);
        return 'ok';
      case 'fill':
        await runAB('fill', action.ref, action.value || '', ...(headed ? ['--headed'] : []));
        await sleep(500);
        return 'ok';
      case 'type':
        await runAB('type', action.ref, action.value || '', ...(headed ? ['--headed'] : []));
        await sleep(300);
        return 'ok';
      case 'press':
        await runAB('press', action.key || 'Enter', ...(headed ? ['--headed'] : []));
        await sleep(1000);
        return 'ok';
      case 'scroll':
        await runAB('scroll', action.direction || 'down', String(action.amount || 500), ...(headed ? ['--headed'] : []));
        await sleep(500);
        return 'ok';
      case 'navigate':
        await runAB('open', action.url, ...(headed ? ['--headed'] : []));
        await sleep(2000);
        return 'ok';
      case 'webmcp':
        console.log(`   🔌 Calling WebMCP tool: ${action.tool} with args: ${JSON.stringify(action.args)}`);
        const result = await webmcpCall(action.tool, action.args || {});
        console.log(`   🔌 WebMCP result: ${JSON.stringify(result).slice(0, 200)}`);
        await sleep(1500);
        return 'ok';
      case 'done':
        console.log(`\n✅ TASK COMPLETE: ${action.result || 'Done'}`);
        return 'done';
      case 'stuck':
        console.log(`\n❌ STUCK: ${action.reason || 'Unknown'}`);
        return 'stuck';
      default:
        console.log(`\n⚠️ Unknown: ${act}`);
        return 'unknown';
    }
  } catch (error) {
    console.log(`   ⚠️ Action error: ${error.message}`);
    return { error: error.message };
  }
}

async function executeActionAsync(action, headed = false) {
  const act = action.action;
  switch (act) {
    case 'screenshot':
      console.log(`   📸 Taking ${action.annotate ? 'annotated ' : ''}screenshot...`);
      const shot = await takeScreenshot(action.annotate !== false);
      console.log(`   📸 Screenshot result: ${JSON.stringify(shot).slice(0, 200)}`);
      lastScreenshot = shot;
      await sleep(1000);
      return 'ok';
    case 'vision':
      if (!action.image && !lastScreenshot) {
        console.log(`   👁️ Vision: No image provided and no recent screenshot`);
        return 'ok';
      }
      let imageData = action.image;
      if (action.image === 'base64' || !action.image) {
        imageData = lastScreenshot?.base64 || lastScreenshot?.data || lastScreenshot?.result;
      }
      if (!imageData) {
        console.log(`   👁️ Vision: No image data available`);
        return 'ok';
      }
      console.log(`   👁️ Analyzing image with prompt: ${action.prompt?.slice(0, 100)}...`);
      const visionResult = await analyzeImageWithNano(action.prompt, imageData);
      console.log(`   👁️ Vision result: ${JSON.stringify(visionResult).slice(0, 300)}`);
      await sleep(1500);
      return 'ok';
    default:
      return executeAction(action, headed);
  }
}

// ── Multimodal Vision Helpers ──
let lastScreenshot = null;

async function takeScreenshot(annotate = true) {
  const cmdArgs = ['screenshot'];
  if (annotate) cmdArgs.push('--annotate');
  return runAB('screenshot', ...cmdArgs);
}

async function analyzeImageWithNano(prompt, imageBase64) {
  return new Promise((resolve, reject) => {
    const req = http.request({
      hostname: 'localhost',
      port: 8765,
      path: '/v1/analyze-image',
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      timeout: 120000,
    }, (res) => {
      let body = '';
      res.on('data', c => body += c);
      res.on('end', () => {
        try {
          const data = JSON.parse(body);
          resolve(data.result || data);
        } catch {
          resolve({ error: 'Parse error' });
        }
      });
    });
    req.on('error', e => reject(new Error(`Network error: ${e.message}`)));
    req.on('timeout', () => { req.destroy(); reject(new Error('Request timeout')); });
    req.write(JSON.stringify({ image: imageBase64, prompt }));
    req.end();
  });
}

// ── WebMCP Helpers ──
async function webmcpDiscover() {
  return new Promise(resolve => {
    const req = http.request({
      hostname: 'localhost', port: 8765, path: '/v1/webmcp/discover',
      method: 'POST', headers: { 'Content-Type': 'application/json' }
    }, res => {
      let body = ''; res.on('data', c => body += c);
      res.on('end', () => { try { resolve(JSON.parse(body)); } catch { resolve({ tools: [] }); }});
    });
    req.on('error', () => resolve({ tools: [] }));
    req.write(JSON.stringify({ tabId: 1 }));
    req.end();
  });
}

async function webmcpCall(toolName, args) {
  return new Promise(resolve => {
    const req = http.request({
      hostname: 'localhost', port: 8765, path: '/v1/webmcp/call',
      method: 'POST', headers: { 'Content-Type': 'application/json' }
    }, res => {
      let body = ''; res.on('data', c => body += c);
      res.on('end', () => { try { resolve(JSON.parse(body)); } catch { resolve({ error: 'Parse error' }); }});
    });
    req.on('error', e => resolve({ error: e.message }));
    req.write(JSON.stringify({ tabId: 1, toolName, args }));
    req.end();
  });
}

async function discoverWebMcpTools() {
  return [];
}

async function checkBridge() {
  return new Promise(resolve => {
    const req = http.get('http://localhost:8765/health', res => {
      let body = '';
      res.on('data', c => body += c);
      res.on('end', () => {
        try {
          const h = JSON.parse(body);
          console.log(`🏥 Bridge: ${h.status} (model: ${h.model})`);
          resolve(true);
        } catch { resolve(false); }
      });
    });
    req.on('error', () => { console.log('❌ Bridge not running!'); resolve(false); });
  });
}

async function runAgent(task, maxSteps = 20, headed = false) {
  console.log('🌐 Gemini Nano Browser Agent');
  console.log(`   Task: ${task}`);
  console.log(`   Max steps: ${maxSteps}`);
  console.log(`   Browser: ${headed ? 'headed' : 'headless'}`);
  console.log();

  let messages = [
    { role: 'system', content: SYSTEM_PROMPT },
    { role: 'user', content: `Task: ${task}\n\nI'll show you the page. Respond with your next action as JSON.` },
  ];

  for (let step = 1; step <= 20; step++) {
    console.log(`── Step ${step}/20 ──`);

    // Get page snapshot
    let pageInfo;
    try {
      const result = await runAB('snapshot');
      if (result.error) throw new Error(result.error);
      const data = result.data || result;
      const parsed = typeof data === 'string' ? JSON.parse(data) : data;
      const tree = parsed.tree || parsed.snapshot || JSON.stringify(parsed);
      const url = parsed.url || 'unknown';
      const title = parsed.title || 'untitled';
      pageInfo = `URL: ${url}\nTitle: ${title}\n\n${tree}`;
    } catch (error) {
      pageInfo = `Could not get snapshot: ${error.message}`;
    }

    if (pageInfo.startsWith('Could not')) {
      console.log(`   ⚠️ Snapshot failed: ${pageInfo}`);
      if (step === 1) {
        messages.push({ role: 'user', content: 'No page is currently open. Navigate to a URL first.' });
      } else {
        messages.push({ role: 'user', content: `Snapshot failed: ${pageInfo}. Try a different action.` });
      }
    } else {
      const context = pageInfo.length > 4000 ? pageInfo.slice(0, 4000) + '\n... (truncated)' : pageInfo;
      messages.push({ role: 'user', content: `Current page:\n${context}` });
    }

    // Trim history
    if (messages.length > 12) {
      messages = [messages[0], ...messages.slice(-10)];
    }

    console.log('   🤔 Asking Gemini Nano...');
    const response = await geminiChat(messages);
    console.log(`   💬 Response: ${response.slice(0, 150)}...`);

    const action = parseAction(response);
    console.log(`   🎯 Action: ${JSON.stringify(action)}`);

    messages.push({ role: 'assistant', content: response });

    const result = await executeAction(action, false);
    
    // Include action result in conversation for the model to see
    if (result && typeof result === 'object') {
      messages.push({ role: 'user', content: `Action result: ${JSON.stringify(result).slice(0, 1000)}` });
    } else if (result && result !== 'ok' && result !== 'done' && result !== 'stuck') {
      messages.push({ role: 'user', content: `Action result: ${result}` });
    }

    if (result === 'done' || result === 'stuck') break;
  }

  console.log('\n🧹 Cleaning up browser...');
  await runAB('close', '--all');
  console.log('Done.');
}

// ── Main ──
const args = process.argv.slice(2);
if (args.length === 0) {
  console.log('Usage: node agent.js "<task>" [--headed] [--max-steps N]');
  process.exit(1);
}

const task = args[0];
const headed = args.includes('--headed');
const maxStepsIdx = args.indexOf('--max-steps');
const maxSteps = maxStepsIdx >= 0 ? parseInt(args[maxStepsIdx + 1]) || 20 : 20;

checkBridge().then(ok => {
  if (!ok) {
    console.log('Start bridge: cd ~/gemini-nano-demo/bridge && node server.js &');
    process.exit(1);
  }
  runAgent(task, 20, headed).catch(e => {
    console.error('Agent error:', e);
    process.exit(1);
  });
});

export { runAgent, checkBridge };